### PR TITLE
[Sanitizer] add signed-integer-wrap sanitizer

### PR DESCRIPTION
**Pull request by JustinStitt**
**Reasoning**
Clang has a `signed-integer-overflow` sanitizer to catch arithmetic overflow; however, most* of its instrumentation [fails to apply](https://godbolt.org/z/ee41rE8o6) when `-fwrapv` is enabled.

The Linux kernel enables `-fno-strict-overflow` which implies `-fwrapv`. This means we are [currently unable to detect signed-integer wrap-around](https://github.com/KSPP/linux/issues/26). All the while, the root cause of many security vulnerabilities in the Linux kernel is [arithmetic overflow](https://cwe.mitre.org/data/definitions/190.html)

In short, we need a way to instrument signed arithmetic even when the overflow behavior is *technically* defined due to `-fwrapv`. I propose we name the sanitizer "signed-integer-wrap" as it best captures the language semantics at hand.

*\* it seems division and modulo arithmetic is still instrumented with `-fsanitize=signed-integer-overflow` and `-fwrapv`*

**Other Notes**
* The `unsigned-integer-overflow` sanitizer is *probably* named incorrectly because "a computation involving unsigned operands can never overflow" [1](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf). 

* Perhaps this sanitizer does not belong in the **U**ndefined **B**ehavior Sanitizer as wrap-around is not UB (although the instrumentation is exactly the same).

* ~~I tried running clang-format but the diff was huge and beyond just my changes.~~ (fixed in [46ef570](https://github.com/llvm/llvm-project/pull/80089/commits/46ef570dd4d80d9d7c827ae2391f3e2b65444d37))

cc: @nickdesaulniers @kees @nathanchance @bwendling

### CODE REVIEW

**Comment by DavidSpickett** [comment]
> I tried running clang-format but the diff was huge and beyond just my changes.

https://clang.llvm.org/docs/ClangFormat.html#git-integration tells you how to narrow it down to just your changes. Once that's setup you can use the command from the comment the bot left.

Sometimes it will try to format too much, in which case it's fine to ignore it, but looking at what the bot reported here that's not the case.

File: clang/test/CodeGen/integer-wrap.c
Comment (tschuett) [inline clang/test/CodeGen/integer-wrap.c:27]: In some tests the branch on the overflow bit is not visible. It gives you more confidence, if there is always a branch on the overflow bit.

**Comment by nickdesaulniers** [comment]
> I tried running clang-format but the diff was huge and beyond just my changes.

I usually do `git clang-format HEAD~N` where N is the number of commits of mine (N is optional when N == 1).

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c]: Makes sense. Fixed in [4848e38](https://github.com/llvm/llvm-project/pull/80089/commits/4848e38223c4191c2e7d45839c2ec34341597c6c).

File: clang/test/CodeGen/integer-wrap.c
Comment (nickdesaulniers) [inline clang/test/CodeGen/integer-wrap.c:1]: Have you seen `llvm/utils/update_cc_test_checks.py` before?  Honestly, ALL new tests to clang (or llvm) involving codegen should strive to use that (or `update_any_test_checks.py` which I believe will now determine the correct underlying `update_*_test_checks.py` to use).

Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c:1]: I must be doing something wrong because it suggest I remove all my `Check: ...`'s and replcae them with `Check: {{.*}}`

File: clang/test/CodeGen/integer-wrap.c
Comment (nickdesaulniers) [inline clang/test/CodeGen/integer-wrap.c:2]: `CHECK` is the implicit default prefix if otherwise unspecified.  Being explicit here is not necessary and is discouraged (unlike the zen of python).

If the first RUN line and the second run line share most output and only differ slightly, then you should use `--check-prefixes=` (plural) with a common shared prefix and non-shared prefix. Example:

```
// RUN: ... --check-prefixes=CHECK,CHECK-FOO
// RUN: ... -check-prefixes=CHECK,CHECK-BAR
```
The generated check lines from update_cc_test_checks will collapse the common lines to CHECK:, and DTRT for the rest.

Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c:2]: > `CHECK` is the implicit default prefix if otherwise unspecified. Being explicit here is not necessary and is discouraged (unlike the zen of python).

Gotcha, should I rename Check to something else or just drop the --check-prefix entirely?


Also, the non-instrumented vs instrumented version don't have too much in common. Should I still opt for this shared --check-prefixes approach? 

Here's what we're workin' with:

<details>
<summary>clang -cc1 -fwrapv -triple x86_64-apple-darwin -emit-llvm -o - -fsanitize=signed-integer-wrap san.c
</summary>
<br>

```

; ModuleID = 'san.c'
source_filename = "san.c"
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-darwin"

@b = external global i32, align 4
@c = external global i32, align 4
@.src = private unnamed_addr constant [6 x i8] c"san.c\00", align 1
@0 = private unnamed_addr constant { i16, i16, [6 x i8] } { i16 0, i16 11, [6 x i8] c"'int'\00" }
@1 = private unnamed_addr global { { ptr, i32, i32 }, ptr } { { ptr, i32, i32 } { ptr @.src, i32 4, i32 9 }, ptr @0 }
@a = external global i32, align 4

; Function Attrs: noinline nounwind optnone
define void @test_add_overflow() #0 {
entry:
  %0 = load volatile i32, ptr @b, align 4
  %1 = load volatile i32, ptr @c, align 4
  %2 = call { i32, i1 } @llvm.sadd.with.overflow.i32(i32 %0, i32 %1), !nosanitize !2
  %3 = extractvalue { i32, i1 } %2, 0, !nosanitize !2
  %4 = extractvalue { i32, i1 } %2, 1, !nosanitize !2
  %5 = xor i1 %4, true, !nosanitize !2
  br i1 %5, label %cont, label %handler.add_overflow, !prof !3, !nosanitize !2

handler.add_overflow:                             ; preds = %entry
  %6 = zext i32 %0 to i64, !nosanitize !2
  %7 = zext i32 %1 to i64, !nosanitize !2
  call void @__ubsan_handle_add_overflow_abort(ptr @1, i64 %6, i64 %7) #3, !nosanitize !2
  unreachable, !nosanitize !2

cont:                                             ; preds = %entry
  store volatile i32 %3, ptr @a, align 4
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
declare { i32, i1 } @llvm.sadd.with.overflow.i32(i32, i32) #1

; Function Attrs: noreturn nounwind uwtable
declare void @__ubsan_handle_add_overflow_abort(ptr, i64, i64) #2

attributes #0 = { noinline nounwind optnone "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-features"="+cx8,+mmx,+sse,+sse2,+x87" }
attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
attributes #2 = { noreturn nounwind uwtable }
attributes #3 = { noreturn nounwind }

!llvm.module.flags = !{!0}
!llvm.ident = !{!1}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{!"clang version 19.0.0git (https://github.com/JustinStitt/llvm-project.git 3766054a825a7dd4ae69feb73c767029f86f912e)"}
!2 = !{}
!3 = !{!"branch_weights", i32 1048575, i32 1}

```
</details>

<details>
<summary>clang -cc1 -fwrapv -triple x86_64-apple-darwin -emit-llvm -o - -fsanitize=signed-integer-overflow san.c

</summary>
<br>

```

; ModuleID = 'san.c'
source_filename = "san.c"
target datalayout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-darwin"

@b = external global i32, align 4
@c = external global i32, align 4
@a = external global i32, align 4

; Function Attrs: noinline nounwind optnone
define void @test_add_overflow() #0 {
entry:
  %0 = load volatile i32, ptr @b, align 4
  %1 = load volatile i32, ptr @c, align 4
  %add = add i32 %0, %1
  store volatile i32 %add, ptr @a, align 4
  ret void
}

attributes #0 = { noinline nounwind optnone "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-features"="+cx8,+mmx,+sse,+sse2,+x87" }

!llvm.module.flags = !{!0}
!llvm.ident = !{!1}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{!"clang version 19.0.0git (https://github.com/JustinStitt/llvm-project.git 3766054a825a7dd4ae69feb73c767029f86f912e)"}

```

</details>

<details open>
<summary>diff</summary>
<br>


```diff

8,10d7
< @.src = private unnamed_addr constant [6 x i8] c"san.c\00", align 1
< @0 = private unnamed_addr constant { i16, i16, [6 x i8] } { i16 0, i16 11, [6 x i8] c"'int'\00" }
< @1 = private unnamed_addr global { { ptr, i32, i32 }, ptr } { { ptr, i32, i32 } { ptr @.src, i32 4, i32 9 }, ptr @0 }
18,31c15,16
<   %2 = call { i32, i1 } @llvm.sadd.with.overflow.i32(i32 %0, i32 %1), !nosanitize !2
<   %3 = extractvalue { i32, i1 } %2, 0, !nosanitize !2
<   %4 = extractvalue { i32, i1 } %2, 1, !nosanitize !2
<   %5 = xor i1 %4, true, !nosanitize !2
<   br i1 %5, label %cont, label %handler.add_overflow, !prof !3, !nosanitize !2
< 
< handler.add_overflow:                             ; preds = %entry
<   %6 = zext i32 %0 to i64, !nosanitize !2
<   %7 = zext i32 %1 to i64, !nosanitize !2
<   call void @__ubsan_handle_add_overflow_abort(ptr @1, i64 %6, i64 %7) #3, !nosanitize !2
<   unreachable, !nosanitize !2
< 
< cont:                                             ; preds = %entry
<   store volatile i32 %3, ptr @a, align 4
---
>   %add = add i32 %0, %1
>   store volatile i32 %add, ptr @a, align 4
35,40d19
< ; Function Attrs: nocallback nofree nosync nounwind speculatable willreturn memory(none)
< declare { i32, i1 } @llvm.sadd.with.overflow.i32(i32, i32) #1
< 
< ; Function Attrs: noreturn nounwind uwtable
< declare void @__ubsan_handle_add_overflow_abort(ptr, i64, i64) #2
< 
42,44d20
< attributes #1 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
< attributes #2 = { noreturn nounwind uwtable }
< attributes #3 = { noreturn nounwind }
51,52d26
< !2 = !{}
< !3 = !{!"branch_weights", i32 1048575, i32 1}


```

</details>

Comment (nickdesaulniers) [reply clang/test/CodeGen/integer-wrap.c:2]: If there is nothing or not much in common, then two distinct `--check-prefix=` (singular) is the way to go.  Avoid using `CHECK` as the single `--check-prefix` though, since that clashes with the implicit default.  i.e. prefer:
```
RUN: ... --check-prefix=CHECK-FOO
RUN: ... --check-prefix=CHECK-BAR
```
to:
```
RUN: ... --check-prefix=CHECK
RUN: ... --check-prefix=CHECK-BAR
```

**Comment by JustinStitt** [comment]
Hey, does anyone know why the CI bot (buildkite) fails my builds for a trailing whitespace in a file I did not touch?

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c]: Gotcha. Fixed in [f790b65](https://github.com/llvm/llvm-project/pull/80089/commits/f790b6522166cc59d900400f9e7beee5d8a19750). Thanks Nick!

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c:9]: I believe @tschuett was saying that to be confident we are instrumenting the arithmetic (and the correct operands therein) we should check the overflow bit usage in the branch instruction. The only way I could think of doing this was capturing the call to the overflow intrinsic which returns the value and the overflow bit. Is there a way to minimize my CHECKs?

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c:1]: I believe @tschuett was saying that to be confident we are instrumenting the arithmetic (and the correct operands therein) we should check the overflow bit usage in the branch instruction. The only way I could think of doing this was capturing the call to the overflow intrinsic which returns the value and the overflow bit. Is there a way to minimize my CHECKs?

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c:9]: I was thinking to be certain that no instrumentation exists we just need to check that no call to the handler exists. The idea being that it's easier to disprove something than to prove it.

**Comment by kees** [comment]
> Hey, does anyone know why the CI bot (buildkite) fails my builds for a trailing whitespace in a file I did not touch?
> 
> It says there's some trailing whitespace in some documentation files that my PR doesn't touch (unless I'm misinterpreting its output): https://buildkite.com/llvm-project/clang-ci/builds/11179#018d6167-a058-43d4-9d2e-6dbd9410a41a

That's mostly just bad luck. It's been fixed now, though, in commit 0217d2e089afba8ca33330713787521ba52a1056 so a it should pass correctly now.

**Comment by JustinStitt** [comment]
ping!

**Comment by kees** [comment]
I've tested this with the Linux kernel, and it is working as expected. It is ready and waiting for this option to gain back a bunch of sanitizer coverage for CIs and likely in production kernels. :)

File: clang/lib/CodeGen/CGExprScalar.cpp
Comment (nickdesaulniers) [inline clang/lib/CodeGen/CGExprScalar.cpp:3579]: Should you be using `hasOneOf` here rather than `||`?

File: clang/test/CodeGen/integer-wrap.c
Comment (nickdesaulniers) [inline clang/test/CodeGen/integer-wrap.c:5]: Wait a second, why are you testing `signed-integer-overflow`? This is already tested in clang/test/codegen/integer-overflow.c. I didn't think your patch is changing the behavior of sio, only adding support for siw?

File: clang/test/CodeGen/integer-wrap.c
Comment (nickdesaulniers) [reply clang/test/CodeGen/integer-wrap.c:1]: > I must be doing something wrong because it suggests I remove all my Check: ...'s and replace them with Check: {{.*}}

Huh, indeed. I can repro. Not sure how to fix, but in digging around I found clang/test/codegen/integer-overflow.c.  Your newly added test file clang/test/CodeGen/integer-wrap.c is highly redundant of that. Why don't you remove clang/test/CodeGen/integer-wrap.c and simply add another RUN line to clang/test/codegen/integer-overflow.c?

**Review by nickdesaulniers — COMMENTED** [review]
PATCH mostly LGTM

UBSan is documented clang/docs/UndefinedBehaviorSanitizer.rst https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html

Please make sure to update the list [of available checks](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#available-checks).

Please also update the release notes to mention this feature.
clang/docs/ReleaseNotes.rst

> The kernel enables -fno-strict-overflow which implies -fwrapv.

While Linux is the only kernel I've contributed to, it's important to be explicit as to which kernel you're referring to. 😛 

> and likely in production kernels. :)

I doubt you meant running in production. 😛

**Comment by kees** [comment]
> > and likely in production kernels. :)
> 
> I doubt you meant running in production. 😛

I very much do -- I expect to use this like we use the array-bounds sanitizer, which is on in production in at least Ubuntu and Android. It may be a long road to getting sane coverage without endless false positives, of course. But the intent is to use it in production.

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c:5]: My patch does not touch SIO but in my testing I wanted it to be clear that SIO and SIW behave differently when `-fwrapv` is enabled. The tests more or less show this difference .

File: clang/lib/CodeGen/CGExprScalar.cpp
Comment (JustinStitt) [reply clang/lib/CodeGen/CGExprScalar.cpp]: Yes, fixed in [e964431](https://github.com/llvm/llvm-project/pull/80089/commits/e96443184c6059ab8c5b97badc9ecc7270d4d466)

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c]: I moved the tests in [5497e8b](https://github.com/llvm/llvm-project/pull/80089/commits/5497e8bc6849bf64c1158ff16b4aa04fd9141920); as such, I am no longer adding tests related to SIO.

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c]: Resolved in [5497e8b](https://github.com/llvm/llvm-project/pull/80089/commits/5497e8bc6849bf64c1158ff16b4aa04fd9141920) since the tests have been moved there is no issue with LABEL statements or prefixes.

File: clang/test/CodeGen/integer-wrap.c
Comment (JustinStitt) [reply clang/test/CodeGen/integer-wrap.c]: Gotcha, I did just that in [5497e8b](https://github.com/llvm/llvm-project/pull/80089/commits/5497e8bc6849bf64c1158ff16b4aa04fd9141920). Thanks Nick!

**Comment by JustinStitt** [comment]
> UBSan is documented clang/docs/UndefinedBehaviorSanitizer.rst https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html
> 
> Please make sure to update the list [of available checks](https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html#available-checks).
> 
> Please also update the release notes to mention this feature. clang/docs/ReleaseNotes.rst
>

Gotcha, done in [e15b6e6](https://github.com/llvm/llvm-project/pull/80089/commits/e15b6e636c27d00a4ef8d0a0f8f648bdcf11d44b) and [b978be6](https://github.com/llvm/llvm-project/pull/80089/commits/b978be62ca2d841d82924b702e8ab2b865754227).

**Review by nickdesaulniers — APPROVED** [review]
I think this patch is good to go. Thanks for working on it!

cc @AaronBallman @efriedma-quic @rjmccall @nikic in case there's any last minute concerns.  I'm happy to sign off on this, but consider giving folks some time to speak up before landing this.

**Comment by JustinStitt** [comment]
> I think this patch is good to go. Thanks for working on it!

Thanks for the review Nick! BTW, I forgot to update the `ReleaseNotes.rst`, I've done so now (hopefully not voiding your previous review)

**Comment by AaronBallman** [comment]
> I think this patch is good to go. Thanks for working on it!
> 
> cc @AaronBallman @efriedma-quic @rjmccall @nikic in case there's any last minute concerns. I'm happy to sign off on this, but consider giving folks some time to speak up before landing this.

No concerns from me, thanks!

**Review by rjmccall — COMMENTED** [review]
Yeah, LGTM.

**Comment by vitalybuka** [comment]
Why not just enforce -fsanitize=signed-integer-overflow with -fwrapv? I suspect it's just overlook, and not intentional behavior.

**Comment by kees** [comment]
> > Why not just enforce -fsanitize=signed-integer-overflow with -fwrapv? I suspect it's just overlook, and not intentional behavior.
> 
> +1
> 
> We should consider this direction

The UB-vs-non-UB seemed to be a really specific goal in the existing code. i.e. that the sanitizer was disabled didn't look like an accident. For people using this to find _only_ UB, this would be a behavioral change, so to me it seems like a separate name makes the most sense. Anyone wanting wrap-around checking can use -wrap, and anyone wanting UB checking can use -overflow.

**Comment by vitalybuka** [comment]
> > > Why not just enforce -fsanitize=signed-integer-overflow with -fwrapv? I suspect it's just overlook, and not intentional behavior.
> > 
> > 
> > +1
> > We should consider this direction
> 
> The UB-vs-non-UB seemed to be a really specific goal in the existing code. i.e. that the sanitizer was disabled didn't look like an accident. For people using this to find _only_ UB, this would be a behavioral change, so to me it seems like a separate name makes the most sense. Anyone wanting wrap-around checking can use -wrap, and anyone wanting UB checking can use -overflow.

Isn't this still UB even with -fwrapv? UB is a language feature, not compiler.

**Comment by vitalybuka** [comment]
> `signed-integer-overflow`

Or maybe I am missing the bigger picture? Is there a plan fix signed-integer-overflow for -fwrapv as well?

**Comment by rjmccall** [comment]
> > > > Why not just enforce -fsanitize=signed-integer-overflow with -fwrapv? I suspect it's just overlook, and not intentional behavior.
> > > 
> > > 
> > > +1
> > > We should consider this direction
> > 
> > 
> > The UB-vs-non-UB seemed to be a really specific goal in the existing code. i.e. that the sanitizer was disabled didn't look like an accident. For people using this to find _only_ UB, this would be a behavioral change, so to me it seems like a separate name makes the most sense. Anyone wanting wrap-around checking can use -wrap, and anyone wanting UB checking can use -overflow.
> 
> Isn't this still UB even with -fwrapv? UB is a language feature, not compiler.

`-fwrapv` is essentially a language dialect that defines the behavior of integer wraparound.  It is no longer UB in compilations using that mode.

**Comment by kees** [comment]
> > > > > Why not just enforce -fsanitize=signed-integer-overflow with -fwrapv? I suspect it's just overlook, and not intentional behavior.
> > > > 
> > > > 
> > > > +1
> > > > We should consider this direction
> > > 
> > > 
> > > The UB-vs-non-UB seemed to be a really specific goal in the existing code. i.e. that the sanitizer was disabled didn't look like an accident. For people using this to find _only_ UB, this would be a behavioral change, so to me it seems like a separate name makes the most sense. Anyone wanting wrap-around checking can use -wrap, and anyone wanting UB checking can use -overflow.
> > 
> > 
> > Isn't this still UB even with -fwrapv? UB is a language feature, not compiler.
> 
> `-fwrapv` is essentially a language dialect that defines the behavior of integer wraparound. It is no longer UB in compilations using that mode.

Right. `-fwrapv` defines the signed integer overflow resolution strategy. Without `-fwrapv` it is undefined (default language feature). With `-fwrapv` it is defined as 2s-complement wrap-around (and is well defined, like unsigned integer overflow).

**Comment by MaskRay** [comment]
This is a UI discussion about how command line options should behave.
Some folks prefer simpler rules while some prefer smart rules (guessing what the user intends).

A [-fwrapv](https://gcc.gnu.org/onlinedocs/gcc/Code-Gen-Options.html#index-fwrapv) user may either:

* rely on the wraparound behavior
* or prevent certain optimizations that would raise security concerns

Our -fsanitize=signed-integer-overflow design have been assuming that -fwrapv users don't need the check.
This PR suggests that an important user does want overflow checks.
It seems very confusing to have two options doing the same thing.

I think we can try -fsanitize=signed-integer-overflow effective when -fwrapv.

There is a precedent that -fsanitize=undefined enables different checks for different targets.
We could make -fsanitize=undefined not imply -fsanitize=signed-integer-overflow when -fwrapv is specified, if we do want to guess the user intention.
Personally I'd prefer moving away from such behaviors and be more orthogonal.

**Comment by kees** [comment]
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=102317

**Comment by kees** [comment]
> Sure -fwrapv makes wraparound defined, but it doesn't prevent us from making -fsanitize=signed-integer-overflow useful. "-fwrapv => no signed-integer-overflow" is not a solid argument.
> 
> I think we can try making -fsanitize=signed-integer-overflow effective even when -fwrapv if specified. -fsanitize=signed-integer-overflow is rare in the wild, probably rarer when combined with -fwrapv.
.

In earlier GCC discussions, it seemed very much like the `-fsanitize=signed-integer-overflow` was meant for UB only, but maybe I misunderstood. See replies leading up to this:
https://gcc.gnu.org/pipermail/gcc-patches/2023-September/630578.html

**Comment by JustinStitt** [comment]
My original idea was to get the SIO sanitizer working with `-fwrapv`, the issue [here](https://github.com/KSPP/linux/issues/26) even suggests it as a viable option. However, after seeing literal checks like:

```cpp
case LangOptions::SOB_Undefined:
        if (!CGF.SanOpts.has(SanitizerKind::SignedIntegerOverflow))
```

... I think the best option is to instrument a wrap sanitizer (which this PR does). A wrap sanitizer best captures the language semantics at hand while maintaining existing functionality of the SIO sanitizer. I think Kees can speak first hand about how picky some folks are about the language being used for this arithmetic overflow/wraparound stuff (he linked some gcc threads above my comment but I've also seen some spicy LKML discussions about how OVERFLOW doesn't exist in the Linux Kernel and as such WRAPAROUND instrumentation is needed).

I think this PR bridges the gap between folks like Kees (who just want all this suspicious kernel arithmetic to go away) and folks like Linus (who is really particular about language).

**Comment by kees** [comment]
GCC folks have not answered. Adding -wrap keeps the behavior for -overflow the same between GCC and Clang. Can we please move this forward and land it as is? We can trivially change this in the future if we need to.

**Review by MaskRay — CHANGES_REQUESTED** [review]
.

**Comment by JustinStitt** [comment]
> I believe we can move forward by reusing `-fsanitize=signed-integer-overflow`, which adds least complexity to Clang and is very reasonable.

I see a few problems with changing `-fsanitize=signed-integer-overflow`:

1) Clang no longer matches GCC's SIO functionality

2) Existing codebases (albeit, very few) may be affected by what is essentially a breaking change. 

3) The C spec language semantics all hint at `signed-integer-overflow` being a misnomer for the properties of the arithmetic we are sanitizing when `-fwrapv` is enabled. (really, `unsigned-integer-overflow` is also poorly named as unsigned arithmetic can't overflow [6.2.5.9](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf))

Should a compiler not at least put some effort into properly representing the semantics at hand?

@MaskRay, there's been lots of good review on this PR with folks more or less liking the direction of it. I'd like to find some common ground on this so we can move it forward. If you really think changing the SIO sanitizer is the way to go I'll probably close this PR and open a new one as it represents a wholly different idea.

File: clang/docs/UndefinedBehaviorSanitizer.rst
Comment (vitalybuka) [inline clang/docs/UndefinedBehaviorSanitizer.rst:203]: Why? division looks like overflow to me?

Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:203]: I was going off the spec, but I think I may have misinterpreted it:

`
An implementation that defines signed integer types as also being modulo need not detect integer overflow, in which case, only integer divide-by-zero need be detected.
`

From [H.2.2 Integer Types](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf)

I initially read this as meaning we need to instrument division no matter what (just in case it's divide by zero or similar case). I can add a check for the signed overflow behavior for the division steps.

Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:203]: > Actually it's very inconsistent that the sanitizer is less strict without `-fwrapv`, when for `signed-integer-overflow` we have an opposite.

Yes, this proposed "wrap" wrap sanitize is less strict when nothing is defined as wrapping (i.e: missing `-fwrapv`). Should this not be the case?


In the same way, `signed-integer-overflow` is more strict without `-fwrapv` as things are  overflowing and not wrapping.

File: clang/docs/UndefinedBehaviorSanitizer.rst
Comment (JustinStitt) [reply clang/docs/UndefinedBehaviorSanitizer.rst:199]: I agree that the current version is not so clear. I must admit, though, I am not 100% sure what changes you would like. Vitaly, do you think it should be worded like such:

```
Note that ``-fwrapv`` implicitly disables most checks performed by this sanitizer 
as the overflow behavior is well-defined.
```

If this doesn't match what you were thinking could you let me know?

**Comment by MaskRay** [comment]
> > I believe we can move forward by reusing `-fsanitize=signed-integer-overflow`, which adds least complexity to Clang and is very reasonable.
> 
> I see a few problems with changing `-fsanitize=signed-integer-overflow`:
> 
> 1. Clang no longer matches GCC's SIO functionality
> 2. Existing codebases (albeit, very few) may be affected by what is essentially a breaking change.
> 
> Should a compiler not at least put some effort into properly representing the semantics at hand?

While I understand adding a new way (to do similar things instead of improving the existing feature) can be tempting, as it avoids "disrupting" current users,
it can lead to software bloat and increased complexity for everyone.
There would be more options for users to read,. Enhancement to one mode might not be ported to the other (similar) one.

I feel that "not breaking existing users" is sometimes exaggerated to argue for not improving an existing feature.
Frankly, we provide ways for users to unblock them. And in this case it seems quite trivial. Add `-fno-sanitize=signed-integer-overflow` or drop the positive option.

I noticed in CodeSearch that this BUILD.gn seems used quite a bit. The author seems aware that with overflow checks, -fwrapv would not make sense.
```
+    if (is_clang && !is_ubsan && !is_ubsan_security) {
+      cflags += [ "-fwrapv" ]
+    }
```

> 3. The C spec language semantics all hint at `signed-integer-overflow` being a misnomer for the properties of the arithmetic we are sanitizing when `-fwrapv` is enabled. (really, `unsigned-integer-overflow` is also poorly named as unsigned arithmetic can't overflow [6.2.5.9](https://www.open-std.org/jtc1/sc22/wg14/www/docs/n1124.pdf))

If there are semantics improvement, unrelated to "let not break users for -fwrapv", I think a new mode is still worth considering.
I haven't read this patch closely, but I believe the patch's intention is just to introduce a new mode working with -fwrapv for the Linux kernel, at the cost of Clang complexity.

> @MaskRay, there's been lots of good review on this PR with folks more or less liking the direction of it. I'd like to find some common ground on this so we can move it forward. If you really think changing the SIO sanitizer is the way to go I'll probably close this PR and open a new one as it represents a wholly different idea.

I agree that a new PR could be useful.

**Comment by kees** [comment]
With PR #82432 landed, this PR is redundant. Thanks for changing the option name! Closing...

**Pull request closed without merging**
